borg/platform_windows.pyx

```diff
@@ -0,0 +1,291 @@
+    ctypedef DWORD* LPDWORD
+    ctypedef int BOOL
+    ctypedef BYTE* PSID
+    struct _ACL:
+        uint16_t AceCount
+
+    HLOCAL LocalFree(HLOCAL)
+    DWORD GetLastError();
+    void SetLastError(DWORD)
+
+    BOOL InitializeSecurityDescriptor(BYTE*, DWORD)
+
+    BOOL LookupAccountNameW(LPCTSTR, LPCTSTR, PSID, LPDWORD, LPCTSTR, LPDWORD, LPDWORD)
+    BOOL GetSecurityDescriptorDacl(PSID, BOOL*, _ACL**, BOOL*)
+
+    cdef extern int ERROR_INSUFFICIENT_BUFFER
+    cdef extern int ERROR_INVALID_SID
+
+    cdef extern int OWNER_SECURITY_INFORMATION
+    cdef extern int GROUP_SECURITY_INFORMATION
+    cdef extern int DACL_SECURITY_INFORMATION
+    cdef extern int SACL_SECURITY_INFORMATION
+    cdef extern int LABEL_SECURITY_INFORMATION
+    cdef extern int ATTRIBUTE_SECURITY_INFORMATION
+    cdef extern int SCOPE_SECURITY_INFORMATION
+    cdef extern int BACKUP_SECURITY_INFORMATION
+    cdef extern int UNPROTECTED_SACL_SECURITY_INFORMATION
+    cdef extern int UNPROTECTED_DACL_SECURITY_INFORMATION
+    cdef extern int PROTECTED_SACL_SECURITY_INFORMATION
+    cdef extern int PROTECTED_DACL_SECURITY_INFORMATION
+
+    cdef extern int SECURITY_DESCRIPTOR_MIN_LENGTH
+
+cdef extern from 'accctrl.h':
+    ctypedef enum _SE_OBJECT_TYPE:
+        SE_FILE_OBJECT
+    ctypedef _SE_OBJECT_TYPE SE_OBJECT_TYPE
+    struct _TRUSTEE_W:
+        uint16_t TrusteeForm
+        uint16_t TrusteeType
+        LPCTSTR ptstrName
+
+    struct _EXPLICIT_ACCESS_W:
+        DWORD grfAccessPermissions
+        uint16_t grfAccessMode
+        DWORD grfInheritance
+        _TRUSTEE_W Trustee
+
+    cdef extern uint16_t TRUSTEE_IS_SID
+    cdef extern uint16_t TRUSTEE_IS_NAME
+    cdef extern uint16_t TRUSTEE_BAD_FORM
+
+    DWORD GetExplicitEntriesFromAclW(_ACL*, uint32_t*, _EXPLICIT_ACCESS_W**)
+
+cdef extern from 'Sddl.h':
+    ctypedef int* LPBOOL
+
+    BOOL GetFileSecurityW(LPCTSTR, int, PSID, DWORD, LPDWORD)
+    BOOL GetSecurityDescriptorOwner(PSID, PSID*, LPBOOL)
+    BOOL LookupAccountSidW(LPCTSTR, PSID, LPCTSTR, LPDWORD, LPCTSTR, LPDWORD, uint16_t*)
+    BOOL ConvertSidToStringSidW(PSID, LPCTSTR*)
+    BOOL ConvertStringSidToSidW(LPCTSTR, PSID*)
+    BOOL ConvertSecurityDescriptorToStringSecurityDescriptorW(BYTE*, DWORD, int, LPCTSTR*, int*)
+
+    cdef extern int SDDL_REVISION_1
+
+cdef extern from 'Aclapi.h':
+    ctypedef void* PACL
+    DWORD GetNamedSecurityInfoW(LPCTSTR, SE_OBJECT_TYPE, DWORD, PSID*, PSID*, PACL*, PACL*, _ACL**)
+    DWORD SetNamedSecurityInfoW(LPCTSTR, int, int, PSID, PSID, PACL, PACL)
+    DWORD SetEntriesInAclW(unsigned int, _EXPLICIT_ACCESS_W*, PACL, _ACL**)
+
+cdef PSID _get_file_security(filename, int request):
+    cdef DWORD length = 0
+    # N.B. This query may fail with ERROR_INVALID_FUNCTION
+    # for some filesystems.
+    cdef wchar_t* wcharfilename = PyUnicode_AsWideCharString(filename, NULL)
+    if wcharfilename == NULL:
+        print 'filename is NULL'
+    GetFileSecurityW(wcharfilename, request, NULL, 0, &length)
```

Comment:
if filename is NULL it should return? Since PyUnicode_AsWideCharString will already raise MemoryError no need to do extra error handling; as soon as control returns to the Python interpreter (after leaving the cdef) the exception will be propagated.